### ISSUE TITLE
Move formatters into plugins

### ISSUE THREAD

**Issue by ilyavolodin** [eslint/eslint#3013]
Should we add formatters to plugins? That would allow formatters to be shared more easily between repository. We also converted everything that used to be file based into plugins, formatters are the only leftover.

Problem with this is how do you specify formatter that's part of the plugin? Currently we deal with plugins in .eslintrc file, but that wouldn't work for formatters, since there could be multiple .eslintrc files and there could only be one formatter.

**Comment by mathieumg** [comment]
:+1: 

I think the formatter used should be the one specified the highest in the ".eslintrc inheritance chain", so either the topmost or the one in the same file that has `root: true`. Otherwise, there could be "conflicts" as you pointed out. Sadly, no formatter specified means to use the default formatter, so we couldn't just get the highest common level with a formatter specified, we'd really need to get the highlest level regardless and use the formatter specified therein, or the default one otherwise.

Another way would be to allow multiple formatters since they're applied on a per-result basis, but that would certainly feel odd visually.

**Comment by royriojas** [comment]
+1 plugins seems to be the way to go. And maybe the formatter we honor for formatters is the one closer to the `process.cwd()`

**Comment by dozoisch** [comment]
:+1: that'd be awesome

**Comment by ilyavolodin** [comment]
I definitely don't want to get into situation where you will have multiple formatters in the output. That just feels completely wrong to me.
I'm also not sold on last formatter in .eslintrc inheritance chain wins approach, because it might cause a lot of confusion and would be very hard to trace in some cases.
Maybe we should still keep the same syntax as we have now (as in CLI) option, but allow `eslint-plugin-formatter/my-awesome-fromatter` to be passed as parameter on CLI?

**Comment by mathieumg** [comment]
> allow eslint-plugin-formatter/my-awesome-fromatter to be passed as parameter on CLI

How would that solve the problem described in #2989? Wouldn't it just be the same, then?

**Comment by ilyavolodin** [comment]
@mathieumg Well it will solve it in two different ways. In plugins, exported objects don't have to be files. You can have multiple exported objects all in the same file. You could then use `eslint-plugin-formatter/my-awesome-formatter2`, so you could have multiple formatters packaged in a single plugin (and single file).

**Comment by mathieumg** [comment]
Does `eslint-plugin-formatter` have dynamic access to the value given after the `/`? Otherwise, I don't see how you would be able to "provide" options to the formatter. other than exporting all combinations programmatically from within, which is a bit less troublesome than generating files for all combinations, but is just as hacky and incurs the performance hit on the runtime.

**Comment by nzakas** [comment]
I'm not sure plugins are the right approach here. Formatters are only specified on the command line, so there's no real value to providing them in plugins, which are specified in the config file. Plus, Right now you could do:

```
eslint -f ./node_modules/foo/bar.js .
```

So you can already specify a formatter from a different module. I don't see a big advantage by trying to do more than that.

**Comment by royriojas** [comment]
@nzakas could you consider adding an option to just spit to the std output the unformatted results? I guess that will solve all of the issues related to the formatters that way some one can just use that as the input for a custom formatter that does accept extra parameters.
some like this

```bash
eslint --raw ./src/**/*.*.js | formatter --no-color --full-paths
```
Sorry for keep coming back on the issue. I do see the value on keeping things simple in the formatters, so if you say there is no way to do something like that, then it is fine.

Thanks for so useful tool

**Comment by nzakas** [comment]
Yeah, we can definitely add a `json` formatter that just spits out the results in the JSON format. So you'd do:

```
eslint -f json . | formatter --no-color --full-paths
```

Would that work?

**Comment by royriojas** [comment]
Yeah, after I suggested it I realize that a json formatter was the way to go :) That should solve all the problems to pass parameters to the formatters. 

At least for me that's enough. Thank you

**Comment by mathieumg** [comment]
Yes, that would be a sensible approach! That way, formatters aren't "jailed" within ESLint and you have full control over them, they don't need ESLint at all anyway since they work "after the fact" and were never allowed access to any internals to begin with.

Thank you.

**Comment by ilyavolodin** [comment]
So in that case, I will close this issue and open a new one to create json reporter.

**Comment by ilyavolodin** [comment]
Opened JSON formatter issue here: #3036
